File: mm-php/src/io/metamessage/mm/MetaMessage.php
<?php

namespace io\metamessage\mm;

class MetaMessage {
    private function __construct() {}

    /**
     * Encode a PHP object to MM bytes. Fields must be accessible;
     * use {@link MM} attribute on fields or types where the wire
     * type should differ from PHP inference.
     */
    public static function encode($root): array {
        return ReflectMmEncoder::encode($root);
    }

    /**
     * Decode MM bytes into a new instance of $clazz.
     * Only a subset of PHP field types is supported;
     * the tree must be a tagged object at the root.
     */
    public static function decode(array $data, string $clazz) {
        $tree = (new WireDecoder($data))->decode();
        return ReflectMmBinder::bind($tree, $clazz);
    }
}

File: mm-java/src/main/java/io/metamessage/mm/MetaMessage.java
package io.github.metamessage.mm;

/**
 * Public entry points for the Java MM codec (binary compatible with the Go {@code mm} package).
 * Field metadata is expressed with {@link MM} instead of Go struct tags.
 */
public final class MetaMessage {

    private MetaMessage() {}

    /**
     * Encode a Java bean graph to MM bytes. Fields must be accessible (typically public or
     * package-private with same-module access); use {@link MM} on fields or types where the wire
     * type should differ from Java inference.
     */
    public static byte[] encode(Object root) throws ReflectiveOperationException {
        return ReflectMmEncoder.encode(root);
    }

    /**
     * Decode MM bytes into a new instance of {@code clazz}. Only a subset of Java field types is
     * supported; the tree must be a tagged object at the root.
     */
    public static <T> T decode(byte[] data, Class<T> clazz) throws Exception {
        MmTree tree = new WireDecoder(data).decode();
        return ReflectMmBinder.bind(tree, clazz);
    }
}
